### PR TITLE
Add validations in parsing the subscriptions

### PR DESCRIPTION
**Pull request by crbrox**
@fgalan 
@kzangeli

### CODE REVIEW

File: src/lib/jsonParseV2/parseSubscription.cpp [hunk 1]
```diff
@@ -303,11 +308,14 @@ static std::string parseEntitiesVector(ConnectionInfo* ciP, std::vector<EntID>*
...
+      else if (idOpt.given)
       {
-        idPattern = ".*";
+        if (forbiddenIdCharsV2(idOpt.value.c_str()))
+        {
+          return badInput(ciP, "forbidden characters in subject entities element id");
+        }
+
+        id= idOpt.value;
       }
```
Comment (kzangeli) [inline src/lib/jsonParseV2/parseSubscription.cpp]: Now, what about id: "" and idPattern: "" ?
Perhaps this is taken care of by the forbidden char checks ... ?

File: src/lib/jsonParseV2/parseSubscription.cpp [hunk 2]
```diff
@@ -319,6 +327,12 @@ static std::string parseEntitiesVector(ConnectionInfo* ciP, std::vector<EntID>*
...
       else if (idPatOpt.given)
       {
+        regex_t regex;
+
+        if (regcomp(&regex, idPatOpt.value.c_str(), 0) != 0)
+        {
+          return badInput(ciP, "invalid regexp in subject entities element idPattern");
+        }
         idPattern = idPatOpt.value;
```
Comment (kzangeli) [inline src/lib/jsonParseV2/parseSubscription.cpp]: And here we just might have that same lapsus :-)
If `regex` isn't used after this block, it must be freed.

File: src/lib/jsonParseV2/parseSubscription.cpp [hunk 3]
```diff
@@ -303,11 +308,14 @@ static std::string parseEntitiesVector(ConnectionInfo* ciP, std::vector<EntID>*
...
-      if (id.empty())
+      else if (idOpt.given)
       {
-        idPattern = ".*";
+        if (forbiddenIdCharsV2(idOpt.value.c_str()))
+        {
+          return badInput(ciP, "forbidden characters in subject entities element id");
+        }
+
+        id= idOpt.value;
```
Comment (fgalan) [inline src/lib/jsonParseV2/parseSubscription.cpp]: nit: `id=` -> `id =`

File: src/lib/parse/forbiddenChars.cpp [hunk 4]
```diff
@@ -97,7 +97,11 @@ bool forbiddenIdChars(const std::string& api, const char* s, const char* excepti
   {
     return forbiddenChars(s, exceptions);  // old behavior
   }
+  return forbiddenIdCharsV2(s, exceptions);
+}
 
+bool forbiddenIdCharsV2(const char* s, const char* exceptions)
```
Comment (fgalan) [inline src/lib/parse/forbiddenChars.cpp]: Missing function preamble.

File: src/lib/jsonParseV2/parseSubscription.cpp (hunk 2 above)
Comment (crbrox) [inline src/lib/jsonParseV2/parseSubscription.cpp]: fixed e6ff22c11d524003b15a3279287f448e79037dc8

File: src/lib/jsonParseV2/parseSubscription.cpp (hunk 1 above)
Comment (crbrox) [inline src/lib/jsonParseV2/parseSubscription.cpp]: fixed f37cdca1908e7bdbb1fea3fa062bd601ac0d81d6

**Comment by kzangeli** [comment]
LGTM (after last fixes, indicated by @fgalan)

**Comment by kzangeli** [comment]
Wait

File: src/lib/jsonParseV2/parseSubscription.cpp (hunk 3 above)
Comment (crbrox) [inline src/lib/jsonParseV2/parseSubscription.cpp]: fixed 7e2545c38f1f0903d28bb937da4ca9ffee6457ab

File: src/lib/jsonParseV2/parseSubscription.cpp (hunk 2 above)
Comment (kzangeli) [inline src/lib/jsonParseV2/parseSubscription.cpp]: With the fix, regfree is called also if regcomp fails.
Not sure what happens if regfree is called with a 'non-regex'.
I would propose to put the call fo regfree **after** returning on invalid regex

File: src/lib/parse/forbiddenChars.cpp (hunk 4 above)
Comment (crbrox) [inline src/lib/parse/forbiddenChars.cpp]: fixed 274a4a7fae0221854a4eaf9ac015d5587c0f1551

File: src/lib/jsonParseV2/parseSubscription.cpp (hunk 2 above)
Comment (crbrox) [inline src/lib/jsonParseV2/parseSubscription.cpp]: fixed 90ed6ef7158cd1e10db0859ebfe0e0af5a842302

**Comment by fgalan** [comment]
LGTM (passing the ball to @kzangeli for final LGTM + merge + delete branch)

**Comment by kzangeli** [comment]
LGTM

**Comment by fgalan** [comment]
LGTM (again :))

**Comment by kzangeli** [comment]
LGTM

**Pull request merged**